File: crates/wail-net/src/signaling.rs
use std::collections::HashMap;

use anyhow::Result;
use futures_util::{SinkExt, StreamExt};
use tokio::sync::mpsc;
use tokio_tungstenite::tungstenite::Message;
use tracing::{debug, error, info, warn};

use wail_core::protocol::{SignalMessage, SyncMessage};

/// WebSocket signaling client that connects to the WAIL signaling server,
/// joins a room, and relays sync messages and audio data.
///
/// Owns the outgoing channels; incoming channels are returned separately
/// from `connect_with_options` so they can be consumed independently.
pub struct SignalingClient {
    /// Outgoing control-plane messages (log, metrics).
    pub outgoing_tx: mpsc::UnboundedSender<SignalMessage>,
    /// Outgoing sync messages (broadcast or targeted).
    sync_outgoing_tx: mpsc::UnboundedSender<SyncOutgoing>,
    /// Outgoing binary audio frames.
    audio_outgoing_tx: mpsc::Sender<Vec<u8>>,
    /// When set, the write task suppresses the automatic `leave` message on close.
    suppress_leave_tx: Option<tokio::sync::watch::Sender<bool>>,
}

/// Internal enum for outgoing sync messages.
enum SyncOutgoing {
    Broadcast(SyncMessage),
    To { peer_id: String, msg: SyncMessage },
}

impl SignalingClient {
    /// Suppress the automatic `leave` message when this client is dropped.
    /// Call this before replacing the client during signaling reconnection.
    pub fn suppress_leave_on_close(&self) {
        if let Some(ref tx) = self.suppress_leave_tx {
            let _ = tx.send(true);
        }
    }

    /// Broadcast a sync message to all peers in the room via the server.
    pub fn broadcast_sync(&self, msg: &SyncMessage) {
        let _ = self.sync_outgoing_tx.send(SyncOutgoing::Broadcast(msg.clone()));
    }

    /// Send a sync message to a specific peer via the server.
    pub fn send_sync_to(&self, peer_id: &str, msg: &SyncMessage) {
        let _ = self.sync_outgoing_tx.send(SyncOutgoing::To {
            peer_id: peer_id.to_string(),
            msg: msg.clone(),
        });
    }

    /// Send a binary audio frame to all peers in the room via the server.
    /// Returns `true` if the frame was queued, `false` if the channel was full (frame dropped).
    pub fn send_audio(&self, data: &[u8]) -> bool {
        match self.audio_outgoing_tx.try_send(data.to_vec()) {
            Ok(()) => true,
            Err(_) => {
                warn!("Audio outgoing channel full — frame dropped");
                false
            }
        }
    }
}

/// A public room returned by the signaling server's list endpoint.
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct PublicRoom {
    pub room: String,
    pub created_at: i64,
    pub peer_count: u32,
    pub display_names: Vec<String>,
    #[serde(default)]
    pub bpm: Option<f64>,
}

#[derive(serde::Deserialize)]
struct ListResponse {
    rooms: Vec<PublicRoom>,
}

/// Fetch the list of public rooms from a signaling server.
///
/// Uses the HTTP `/rooms` endpoint (not WebSocket).
pub async fn list_public_rooms(base_url: &str) -> Result<Vec<PublicRoom>> {
    let http_url = base_url
        .replace("wss://", "https://")
        .replace("ws://", "http://");
    let base = http_url.trim_end_matches('/');
    let resp = reqwest::Client::new()
        .get(format!("{base}/rooms"))
        .send()
        .await?
        .error_for_status()?;
    let list: ListResponse = resp.json().await?;
    Ok(list.rooms)
}

/// Server response messages (tagged JSON).
#[derive(serde::Deserialize)]
#[serde(tag = "type")]
enum ServerMsg {
    #[serde(rename = "join_ok")]
    JoinOk {
        peers: Vec<String>,
        #[serde(default)]
        peer_display_names: HashMap<String, Option<String>>,
    },
    #[serde(rename = "join_error")]
    JoinError {
        code: String,
        #[serde(default)]
        min_version: Option<String>,
        #[serde(default)]
        slots_available: Option<u64>,
    },
    #[serde(rename = "peer_joined")]
    PeerJoined {
        peer_id: String,
        display_name: Option<String>,
    },
    #[serde(rename = "peer_left")]
    PeerLeft {
        peer_id: String,
    },
    /// Legacy WebRTC signaling — kept for backward compatibility but ignored.
    #[serde(rename = "signal")]
    #[allow(dead_code)]
    Signal {
        #[serde(default)]
        to: String,
        #[serde(default)]
        from: String,
        #[serde(default)]
        payload: serde_json::Value,
    },
    /// Sync message relayed by the server from another peer.
    #[serde(rename = "sync")]
    Sync {
        from: String,
        payload: serde_json::Value,
    },
    #[serde(rename = "evicted")]
    Evicted,
    #[serde(rename = "log")]
    Log {
        from: String,
        level: String,
        target: String,
        message: String,
        timestamp_us: u64,
    },
}

/// Channels returned by `SignalingClient::connect_with_options` for incoming data.
pub struct SignalingChannels {
    /// Control-plane messages from the server (PeerJoined, PeerLeft, LogBroadcast, etc.)
    pub incoming_rx: mpsc::UnboundedReceiver<SignalMessage>,
    /// Sync messages relayed from remote peers via the server.
    pub sync_rx: mpsc::UnboundedReceiver<(String, SyncMessage)>,
    /// Binary audio frames relayed from remote peers via the server.
    pub audio_rx: mpsc::Receiver<(String, Vec<u8>)>,
}

impl SignalingClient {
    /// Connect to the WebSocket signaling server and join a room.
    pub async fn connect(
        server_url: &str,
        room: &str,
        peer_id: &str,
        password: Option<&str>,
    ) -> Result<(Self, SignalingChannels, HashMap<String, Option<String>>)> {
        Self::connect_with_options(server_url, room, peer_id, password, 1, None).await
    }

    /// Connect with full options including stream count and display name.
    ///
    /// Returns the client (for sending), incoming channels (for receiving),
    /// and initial peer display names from the join response.
    pub async fn connect_with_options(
        server_url: &str,
        room: &str,
        peer_id: &str,
        password: Option<&str>,
        stream_count: u16,
        display_name: Option<&str>,
    ) -> Result<(Self, SignalingChannels, HashMap<String, Option<String>>)> {
        let ws_url = format!("{}/ws", server_url.trim_end_matches('/'));

        let (ws_stream, _) = tokio_tungstenite::connect_async(&ws_url).await?;
        let (mut ws_write, mut ws_read) = ws_stream.split();

        // Send join message
        let mut join_msg = serde_json::json!({
            "type": "join",
            "room": room,
            "peer_id": peer_id,
            "stream_count": stream_count,
            "client_version": env!("CARGO_PKG_VERSION"),
        });
        if let Some(pw) = password {
            join_msg["password"] = serde_json::Value::String(pw.to_string());
        }
        if let Some(name) = display_name {
            join_msg["display_name"] = serde_json::Value::String(name.to_string());
        }
        ws_write
            .send(Message::Text(join_msg.to_string()))
            .await?;

        // Wait for join_ok or join_error
        let join_response = loop {
            match ws_read.next().await {
                Some(Ok(Message::Text(text))) => {
                    break serde_json::from_str::<ServerMsg>(&text)?;
                }
                Some(Ok(Message::Close(_))) | None => {
                    anyhow::bail!("WebSocket closed before join response");
                }
                Some(Err(e)) => {
                    anyhow::bail!("WebSocket error waiting for join response: {e}");
                }
                _ => continue,
            }
        };

        let (peers, initial_peer_names) = match join_response {
            ServerMsg::JoinOk {
                peers,
                peer_display_names,
            } => (peers, peer_display_names),
            ServerMsg::JoinError {
                code,
                min_version,
                slots_available,
            } => match code.as_str() {
                "unauthorized" => {
                    anyhow::bail!(
                        "Invalid room password — the room exists and the password doesn't match"
                    );
                }
                "room_full" => {
                    let slots = slots_available.unwrap_or(0);
                    anyhow::bail!("Room full — only {slots} stream slots available");
                }
                "version_outdated" => {
                    let min = min_version.as_deref().unwrap_or("unknown");
                    anyhow::bail!(
                        "Your WAIL version ({}) is outdated. Please update to at least version {min}.",
                        env!("CARGO_PKG_VERSION")
                    );
                }
                other => anyhow::bail!("Join failed: {other}"),
            },
            _ => anyhow::bail!("Unexpected server message before join_ok"),
        };

        info!(
            %server_url, %room, %peer_id,
            existing_peers = peers.len(),
            "Joined signaling room via WebSocket"
        );

        // Control-plane channel
        let (incoming_tx, incoming_rx) = mpsc::unbounded_channel();
        // Sync messages from remote peers
        let (sync_incoming_tx, sync_rx) = mpsc::unbounded_channel();
        // Audio frames from remote peers
        let (audio_incoming_tx, audio_rx) = mpsc::channel(1024);
        // Outgoing control-plane
        let (outgoing_tx, mut outgoing_rx) = mpsc::unbounded_channel::<SignalMessage>();
        // Outgoing sync
        let (sync_outgoing_tx, mut sync_outgoing_rx) = mpsc::unbounded_channel::<SyncOutgoing>();
        // Outgoing audio
        let (audio_outgoing_tx, mut audio_outgoing_rx) = mpsc::channel::<Vec<u8>>(256);

        let (suppress_leave_tx, suppress_leave_rx) = tokio::sync::watch::channel(false);

        // Push PeerList so PeerMesh sees existing peers
        if incoming_tx
            .send(SignalMessage::PeerList { peers })
            .is_err()
        {
            anyhow::bail!("incoming channel closed immediately");
        }

        // Spawn read task: server → incoming channels
        tokio::spawn(async move {
            while let Some(msg_result) = ws_read.next().await {
                match msg_result {
                    Ok(Message::Binary(data)) => {
                        // Binary frame: [1 byte: peer_id_len][peer_id][audio_payload]
                        if data.is_empty() {
                            continue;
                        }
                        let pid_len = data[0] as usize;
                        if data.len() < 1 + pid_len {
                            warn!("Binary frame too short for sender header");
                            continue;
                        }
                        let peer_id = match std::str::from_utf8(&data[1..1 + pid_len]) {
                            Ok(s) => s.to_string(),
                            Err(_) => {
                                warn!("Invalid UTF-8 in binary sender header");
                                continue;
                            }
                        };
                        let audio_data = data[1 + pid_len..].to_vec();
                        if audio_incoming_tx.send((peer_id, audio_data)).await.is_err() {
                            info!("Audio incoming channel closed, stopping WS read");
                            return;
                        }
                    }
                    Ok(Message::Text(text)) => {
                        match serde_json::from_str::<ServerMsg>(&text) {
                            Ok(server_msg) => {
                                match server_msg {
                                    ServerMsg::PeerJoined {
                                        peer_id,
                                        display_name,
                                    } => {
                                        let _ = incoming_tx.send(SignalMessage::PeerJoined {
                                            peer_id,
                                            display_name,
                                        });
                                    }
                                    ServerMsg::PeerLeft { peer_id } => {
                                        let _ = incoming_tx.send(SignalMessage::PeerLeft { peer_id });
                                    }
                                    ServerMsg::Sync { from, payload } => {
                                        match serde_json::from_value::<SyncMessage>(payload) {
                                            Ok(msg) => {
                                                if sync_incoming_tx.send((from, msg)).is_err() {
                                                    info!("Sync incoming channel closed, stopping WS read");
                                                    return;
                                                }
                                            }
                                            Err(e) => {
                                                warn!(error = %e, "Failed to parse relayed sync payload");
                                            }
                                        }
                                    }
                                    ServerMsg::Evicted => {
                                        warn!("Server evicted us — closing signaling");
                                        return;
                                    }
                                    ServerMsg::Log { from, level, target, message, timestamp_us } => {
                                        let _ = incoming_tx.send(SignalMessage::LogBroadcast { from, level, target, message, timestamp_us });
                                    }
                                    ServerMsg::Signal { .. } => {
                                        debug!("Ignoring legacy signal message");
                                    }
                                    _ => {}
                                }
                            }
                            Err(e) => {
                                warn!(error = %e, body = %text, "Failed to parse server message");
                            }
                        }
                    }
                    Ok(Message::Close(_)) => {
                        info!("WebSocket closed by server");
                        return;
                    }
                    Err(e) => {
                        error!(error = %e, "WebSocket read error");
                        return;
                    }
                    _ => {} // ping/pong handled by tungstenite
                }
            }
            info!("WebSocket stream ended");
        });

        // Spawn write task: outgoing channels → server
        tokio::spawn(async move {
            loop {
                tokio::select! {
                    msg = outgoing_rx.recv() => {
                        let Some(msg) = msg else { break };
                        let raw = match &msg {
                            SignalMessage::LogBroadcast { level, target, message, timestamp_us, .. } => {
                                serde_json::json!({
                                    "type": "log",
                                    "level": level,
                                    "target": target,
                                    "message": message,
                                    "timestamp_us": timestamp_us,
                                })
                            }
                            SignalMessage::MetricsReport { dc_open, plugin_connected, per_peer, ipc_drops, boundary_drift_us } => {
                                serde_json::json!({
                                    "type": "metrics_report",
                                    "dc_open": dc_open,
                                    "plugin_connected": plugin_connected,
                                    "per_peer": per_peer,
                                    "ipc_drops": ipc_drops,
                                    "boundary_drift_us": boundary_drift_us,
                                })
                            }
                            _ => continue,
                        };
                        if ws_write.send(Message::Text(raw.to_string())).await.is_err() {
                            warn!("WebSocket write failed — connection lost");
                            return;
                        }
                    }
                    sync_msg = sync_outgoing_rx.recv() => {
                        let Some(sync_msg) = sync_msg else { break };
                        let raw = match sync_msg {
                            SyncOutgoing::Broadcast(msg) => {
                                serde_json::json!({
                                    "type": "sync",
                                    "payload": serde_json::to_value(&msg).unwrap_or_default(),
                                })
                            }
                            SyncOutgoing::To { peer_id, msg } => {
                                serde_json::json!({
                                    "type": "sync_to",
                                    "to": peer_id,
                                    "payload": serde_json::to_value(&msg).unwrap_or_default(),
                                })
                            }
                        };
                        if ws_write.send(Message::Text(raw.to_string())).await.is_err() {
                            warn!("WebSocket write failed — connection lost");
                            return;
                        }
                    }
                    audio_data = audio_outgoing_rx.recv() => {
                        let Some(data) = audio_data else { break };
                        if ws_write.send(Message::Binary(data)).await.is_err() {
                            warn!("WebSocket write failed (audio) — connection lost");
                            return;
                        }
                    }
                }
            }
            // All outgoing channels closed — only send leave if not suppressed
            if *suppress_leave_rx.borrow() {
                info!("Outgoing channel closed, leave suppressed (reconnecting)");
            } else {
                info!("Outgoing channel closed, sending leave");
                let _ = ws_write
                    .send(Message::Text(
                        serde_json::json!({"type": "leave"}).to_string(),
                    ))
                    .await;
            }
            let _ = ws_write.close().await;
        });

        let channels = SignalingChannels {
            incoming_rx,
            sync_rx,
            audio_rx,
        };

        Ok((
            Self {
                outgoing_tx,
                sync_outgoing_tx,
                audio_outgoing_tx,
                suppress_leave_tx: Some(suppress_leave_tx),
            },
            channels,
            initial_peer_names,
        ))
    }
}

File: wail-app/signaling.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"strings"
	"sync"

	"github.com/gorilla/websocket"
)

const appVersion = "2.0.0-go"

// PublicRoom represents a public room from the signaling server.
type PublicRoom struct {
	Room         string   `json:"room"`
	CreatedAt    int64    `json:"created_at"`
	PeerCount    uint32   `json:"peer_count"`
	DisplayNames []string `json:"display_names"`
	BPM          *float64 `json:"bpm,omitempty"`
}

// ListPublicRooms fetches the list of public rooms from a signaling server.
func ListPublicRooms(baseURL string) ([]PublicRoom, error) {
	httpURL := strings.Replace(baseURL, "wss://", "https://", 1)
	httpURL = strings.Replace(httpURL, "ws://", "http://", 1)
	httpURL = strings.TrimRight(httpURL, "/")

	resp, err := http.Get(httpURL + "/rooms")
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("rooms endpoint returned %d", resp.StatusCode)
	}

	var result struct {
		Rooms []PublicRoom `json:"rooms"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return result.Rooms, nil
}

// SignalingClient manages a WebSocket connection to the signaling server.
type SignalingClient struct {
	mu   sync.Mutex
	conn *websocket.Conn

	// Outgoing channels
	syncOutCh  chan outgoingSync
	audioOutCh chan []byte
	ctrlOutCh  chan SignalMessage

	// Suppress leave on close (for reconnection)
	suppressLeave bool

	cancel context.CancelFunc
}

type outgoingSync struct {
	broadcast bool
	peerID    string
	msg       SyncMessage
}

// SignalingChannels holds the incoming channels from the signaling connection.
type SignalingChannels struct {
	IncomingCh chan SignalMessage                // control plane (PeerJoined, PeerLeft, etc.)
	SyncCh     chan FromPeerSync                // sync messages from peers
	AudioCh    chan FromPeerAudio               // binary audio from peers
}

// FromPeerSync wraps a sync message with the sender's peer ID.
type FromPeerSync struct {
	From string
	Msg  SyncMessage
}

// FromPeerAudio wraps binary audio data with the sender's peer ID.
type FromPeerAudio struct {
	From string
	Data []byte
}

// ConnectSignaling connects to the signaling server and joins a room.
// Returns the client, channels, and initial peer display names.
func ConnectSignaling(
	ctx context.Context,
	serverURL, room, peerID string,
	password *string,
	streamCount uint16,
	displayName *string,
) (*SignalingClient, *SignalingChannels, map[string]*string, error) {
	wsURL := strings.TrimRight(serverURL, "/") + "/ws"

	conn, _, err := websocket.DefaultDialer.DialContext(ctx, wsURL, nil)
	if err != nil {
		return nil, nil, nil, fmt.Errorf("websocket connect: %w", err)
	}

	// Send join message
	joinMsg := map[string]any{
		"type":           "join",
		"room":           room,
		"peer_id":        peerID,
		"stream_count":   streamCount,
		"client_version": appVersion,
	}
	if password != nil {
		joinMsg["password"] = *password
	}
	if displayName != nil {
		joinMsg["display_name"] = *displayName
	}
	if err := conn.WriteJSON(joinMsg); err != nil {
		conn.Close()
		return nil, nil, nil, fmt.Errorf("send join: %w", err)
	}

	// Wait for join_ok or join_error
	var serverMsg ServerMsg
	for {
		_, data, err := conn.ReadMessage()
		if err != nil {
			conn.Close()
			return nil, nil, nil, fmt.Errorf("read join response: %w", err)
		}
		if err := json.Unmarshal(data, &serverMsg); err != nil {
			continue
		}
		if serverMsg.Type == "join_ok" || serverMsg.Type == "join_error" {
			break
		}
	}

	if serverMsg.Type == "join_error" {
		conn.Close()
		switch serverMsg.Code {
		case "unauthorized":
			return nil, nil, nil, fmt.Errorf("invalid room password")
		case "room_full":
			slots := uint64(0)
			if serverMsg.SlotsAvailable != nil {
				slots = *serverMsg.SlotsAvailable
			}
			return nil, nil, nil, fmt.Errorf("room full — only %d slots available", slots)
		case "version_outdated":
			min := "unknown"
			if serverMsg.MinVersion != nil {
				min = *serverMsg.MinVersion
			}
			return nil, nil, nil, fmt.Errorf("version %s outdated, need %s", appVersion, min)
		default:
			return nil, nil, nil, fmt.Errorf("join failed: %s", serverMsg.Code)
		}
	}

	log.Printf("[signaling] Joined room %s as %s (%d existing peers)", room, peerID, len(serverMsg.Peers))

	// Set up channels
	incomingCh := make(chan SignalMessage, 64)
	syncCh := make(chan FromPeerSync, 256)
	audioCh := make(chan FromPeerAudio, 1024)
	syncOutCh := make(chan outgoingSync, 256)
	audioOutCh := make(chan []byte, 256)
	ctrlOutCh := make(chan SignalMessage, 64)

	// Push initial PeerList
	incomingCh <- SignalMessage{Type: "PeerList", Peers: serverMsg.Peers}

	childCtx, cancel := context.WithCancel(ctx)

	client := &SignalingClient{
		conn:       conn,
		syncOutCh:  syncOutCh,
		audioOutCh: audioOutCh,
		ctrlOutCh:  ctrlOutCh,
		cancel:     cancel,
	}

	channels := &SignalingChannels{
		IncomingCh: incomingCh,
		SyncCh:     syncCh,
		AudioCh:    audioCh,
	}

	// Spawn read goroutine
	go func() {
		defer close(incomingCh)
		for {
			msgType, data, err := conn.ReadMessage()
			if err != nil {
				select {
				case <-childCtx.Done():
				default:
					log.Printf("[signaling] WebSocket read error: %v", err)
				}
				return
			}

			if msgType == websocket.BinaryMessage {
				// Binary: [1 byte peer_id_len][peer_id][audio_payload]
				if len(data) < 1 {
					continue
				}
				pidLen := int(data[0])
				if len(data) < 1+pidLen {
					continue
				}
				from := string(data[1 : 1+pidLen])
				audioData := make([]byte, len(data)-(1+pidLen))
				copy(audioData, data[1+pidLen:])
				select {
				case audioCh <- FromPeerAudio{From: from, Data: audioData}:
				default:
					// Drop if channel full
				}
				continue
			}

			// Text message
			var msg ServerMsg
			if err := json.Unmarshal(data, &msg); err != nil {
				log.Printf("[signaling] Failed to parse server message: %v", err)
				continue
			}

			switch msg.Type {
			case "peer_joined":
				incomingCh <- SignalMessage{
					Type:        "PeerJoined",
					PeerID:      msg.PeerID,
					DisplayName: msg.DisplayName,
				}
			case "peer_left":
				incomingCh <- SignalMessage{Type: "PeerLeft", PeerID: msg.PeerID}
			case "sync":
				var syncMsg SyncMessage
				if err := json.Unmarshal(msg.Payload, &syncMsg); err != nil {
					log.Printf("[signaling] Failed to parse sync payload: %v", err)
					continue
				}
				select {
				case syncCh <- FromPeerSync{From: msg.From, Msg: syncMsg}:
				default:
				}
			case "evicted":
				log.Printf("[signaling] Evicted by server")
				return
			case "log":
				incomingCh <- SignalMessage{
					Type:        "LogBroadcast",
					From:        msg.From,
					Level:       msg.Level,
					Target:      msg.Target,
					Message:     msg.Message,
					TimestampUs: msg.TimestampUs,
				}
			}
		}
	}()

	// Spawn write goroutine
	go func() {
		for {
			select {
			case <-childCtx.Done():
				if !client.suppressLeave {
					conn.WriteJSON(map[string]string{"type": "leave"})
				}
				conn.Close()
				return
			case out := <-syncOutCh:
				var raw map[string]any
				payload, _ := json.Marshal(out.msg)
				if out.broadcast {
					raw = map[string]any{
						"type":    "sync",
						"payload": json.RawMessage(payload),
					}
				} else {
					raw = map[string]any{
						"type":    "sync_to",
						"to":      out.peerID,
						"payload": json.RawMessage(payload),
					}
				}
				if err := conn.WriteJSON(raw); err != nil {
					log.Printf("[signaling] Write failed: %v", err)
					return
				}
			case data := <-audioOutCh:
				if err := conn.WriteMessage(websocket.BinaryMessage, data); err != nil {
					log.Printf("[signaling] Audio write failed: %v", err)
					return
				}
			case msg := <-ctrlOutCh:
				var raw map[string]any
				switch msg.Type {
				case "LogBroadcast":
					raw = map[string]any{
						"type":         "log",
						"level":        msg.Level,
						"target":       msg.Target,
						"message":      msg.Message,
						"timestamp_us": msg.TimestampUs,
					}
				case "MetricsReport":
					raw = map[string]any{
						"type":              "metrics_report",
						"dc_open":           msg.DCOpen,
						"plugin_connected":  msg.PluginConnected,
						"per_peer":          msg.PerPeer,
						"ipc_drops":         msg.IPCDrops,
						"boundary_drift_us": msg.BoundaryDriftUs,
					}
				default:
					continue
				}
				if err := conn.WriteJSON(raw); err != nil {
					log.Printf("[signaling] Control write failed: %v", err)
					return
				}
			}
		}
	}()

	return client, channels, serverMsg.PeerDisplayNames, nil
}

// BroadcastSync broadcasts a sync message to all peers.
func (sc *SignalingClient) BroadcastSync(msg SyncMessage) {
	select {
	case sc.syncOutCh <- outgoingSync{broadcast: true, msg: msg}:
	default:
	}
}

// SendSyncTo sends a sync message to a specific peer.
func (sc *SignalingClient) SendSyncTo(peerID string, msg SyncMessage) {
	select {
	case sc.syncOutCh <- outgoingSync{broadcast: false, peerID: peerID, msg: msg}:
	default:
	}
}

// SendAudio broadcasts binary audio data.
func (sc *SignalingClient) SendAudio(data []byte) {
	select {
	case sc.audioOutCh <- data:
	default:
	}
}

// SendControl sends a control-plane message (log, metrics).
func (sc *SignalingClient) SendControl(msg SignalMessage) {
	select {
	case sc.ctrlOutCh <- msg:
	default:
	}
}

// SuppressLeaveOnClose prevents the automatic leave message on disconnect (for reconnection).
func (sc *SignalingClient) SuppressLeaveOnClose() {
	sc.mu.Lock()
	defer sc.mu.Unlock()
	sc.suppressLeave = true
}

// Close shuts down the signaling client.
func (sc *SignalingClient) Close() {
	if sc.cancel != nil {
		sc.cancel()
	}
}

// PeerMesh manages communication with all peers via the signaling server.
type PeerMesh struct {
	peerID           string
	signaling        *SignalingClient
	channels         *SignalingChannels
	peers            map[string]bool
	streamCount      uint16
	initialPeerNames map[string]*string
	mu               sync.Mutex
}

// NewPeerMesh creates a mesh from a signaling client.
func NewPeerMesh(peerID string, signaling *SignalingClient, channels *SignalingChannels, streamCount uint16, initialPeerNames map[string]*string) *PeerMesh {
	return &PeerMesh{
		peerID:           peerID,
		signaling:        signaling,
		channels:         channels,
		peers:            make(map[string]bool),
		streamCount:      streamCount,
		initialPeerNames: initialPeerNames,
	}
}

// Broadcast sends a sync message to all peers.
func (m *PeerMesh) Broadcast(msg SyncMessage) {
	m.signaling.BroadcastSync(msg)
}

// SendTo sends a sync message to a specific peer.
func (m *PeerMesh) SendTo(peerID string, msg SyncMessage) error {
	m.signaling.SendSyncTo(peerID, msg)
	return nil
}

// BroadcastAudio sends binary audio to all peers.
func (m *PeerMesh) BroadcastAudio(data []byte) {
	m.signaling.SendAudio(data)
}

// PollSignaling processes one signaling event and returns a MeshEvent.
func (m *PeerMesh) PollSignaling() (*MeshEvent, bool) {
	msg, ok := <-m.channels.IncomingCh
	if !ok {
		return nil, false
	}
	return m.handleSignalMessage(msg), true
}

func (m *PeerMesh) handleSignalMessage(msg SignalMessage) *MeshEvent {
	m.mu.Lock()
	defer m.mu.Unlock()

	switch msg.Type {
	case "PeerList":
		for _, pid := range msg.Peers {
			if pid != m.peerID {
				m.peers[pid] = true
			}
		}
		return &MeshEvent{Type: "PeerListReceived", PeerCount: len(msg.Peers)}
	case "PeerJoined":
		m.peers[msg.PeerID] = true
		return &MeshEvent{Type: "PeerJoined", PeerID: msg.PeerID, DisplayName: msg.DisplayName}
	case "PeerLeft":
		delete(m.peers, msg.PeerID)
		return &MeshEvent{Type: "PeerLeft", PeerID: msg.PeerID}
	case "LogBroadcast":
		return &MeshEvent{
			Type: "PeerLogBroadcast", From: msg.From,
			Level: msg.Level, Target: msg.Target,
			Message: msg.Message, TimestampUs: msg.TimestampUs,
		}
	}
	return nil
}

// ConnectedPeers returns a list of known peer IDs.
func (m *PeerMesh) ConnectedPeers() []string {
	m.mu.Lock()
	defer m.mu.Unlock()
	peers := make([]string, 0, len(m.peers))
	for pid := range m.peers {
		peers = append(peers, pid)
	}
	return peers
}

// TakeInitialPeerNames returns and clears the initial peer names.
func (m *PeerMesh) TakeInitialPeerNames() map[string]*string {
	m.mu.Lock()
	defer m.mu.Unlock()
	names := m.initialPeerNames
	m.initialPeerNames = nil
	return names
}

// AnyPeersConnected returns true if any peers are known.
func (m *PeerMesh) AnyPeersConnected() bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	return len(m.peers) > 0
}

// IsPeerConnected checks if a specific peer is known.
func (m *PeerMesh) IsPeerConnected(peerID string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.peers[peerID]
}

// RemovePeer removes a peer from tracking.
func (m *PeerMesh) RemovePeer(peerID string) {
	m.mu.Lock()
	defer m.mu.Unlock()
	delete(m.peers, peerID)
}

// SendLog sends a log broadcast to the signaling server.
func (m *PeerMesh) SendLog(level, target, message string, timestampUs uint64) {
	m.signaling.SendControl(SignalMessage{
		Type: "LogBroadcast", Level: level, Target: target,
		Message: message, TimestampUs: timestampUs,
	})
}

// SendMetricsReport sends a metrics report to the signaling server.
func (m *PeerMesh) SendMetricsReport(dcOpen, pluginConnected bool, perPeer map[string]PeerFrameReport, ipcDrops uint64, boundaryDriftUs *int64) {
	m.signaling.SendControl(SignalMessage{
		Type: "MetricsReport", DCOpen: dcOpen, PluginConnected: pluginConnected,
		PerPeer: perPeer, IPCDrops: ipcDrops, BoundaryDriftUs: boundaryDriftUs,
	})
}

// ReconnectSignaling reconnects the WebSocket and returns new channels.
func (m *PeerMesh) ReconnectSignaling(
	ctx context.Context,
	serverURL, room string,
	password *string,
	displayName *string,
) (*SignalingChannels, map[string]*string, error) {
	// Suppress leave on old connection
	m.signaling.SuppressLeaveOnClose()
	m.signaling.Close()

	newClient, newChannels, peerNames, err := ConnectSignaling(
		ctx, serverURL, room, m.peerID, password, m.streamCount, displayName,
	)
	if err != nil {
		return nil, nil, err
	}

	m.mu.Lock()
	m.signaling = newClient
	m.channels = newChannels
	m.initialPeerNames = peerNames
	m.mu.Unlock()

	// Consume initial PeerList
	if msg, ok := <-newChannels.IncomingCh; ok && msg.Type == "PeerList" {
		m.mu.Lock()
		for _, pid := range msg.Peers {
			if pid != m.peerID {
				m.peers[pid] = true
			}
		}
		m.mu.Unlock()
	}

	return newChannels, peerNames, nil
}
